### PR TITLE
[REVIEW] Allow construction of cuda_async_memory_resource from existing pool

### PR DESCRIPTION
**Pull request by fkallen**
Adds a new MR type `cuda_async_view_memory_resource` which has a constructor `cuda_async_view_memory_resource(cudaMemPool_t valid_pool_handle)` . The memory resource will use this pool for allocation and deallocation instead of managing its own pool.

Refactors `cuda_async_memory_resource` to have an instance of the above and create it with a `cudaMemPool_t` that it owns.

### CODE REVIEW

**Comment by harrism** [comment]
Another approach could be to have an owning `cuda_pool` type that is simply a RAII wrapper for `cudaMemPool_t`, and a non-owning `cuda_pool_view` type that can be constructed either from a `cuda_pool` or a `cudaMemPool_t`. Then have `cuda_async_memory_resource` take a `cuda_pool_view` and *never* own the pool.

**Comment by harrism** [comment]
This still feels clunky to me. I think it's mostly because of the name `cuda_pool_wrapper`. The name is vague. To me a pool wrapper sounds like it's not a memory_resource, but just some RAII wrapper for something to be owned. Especially confusing since we have another resource in RMM called an `owning_wrapper'. Memory_resources so far always have `memory_resource` or `resource_adaptor` in the name. `cuda_pool_wrapper` does exactly what `cuda_async_memory_resource` does except it doesn't own the pool.  The only difference between the two resources is one is owning and the other is non-owning.  So I think the naming should reflect that (e.g. shared vs. unique, or somehow make one a view, etc).

**Comment by harrism** [comment]
Due to open discussion, I'm moving this to the next release. @fkallen can you merge rapidsai:branch-22.02 into your branch in order to target the right branch?

**Comment by fkallen** [comment]
At the moment I do not know how to further proceed with this PR. There seem to be two open questions with different opinions.
1. What should be the name of the new non-owning resource?
   `cuda_pool_wrapper`, `cuda_async_non_owning_memory_resource` , `cuda_async_shared_memory_resource`, `cuda_async_view_memory_resource`, ... ?
2. Should the current `cuda_async_memory_resource` be able to take ownership of an existing pool, or should it always create its own pool ?

**Comment by jrhemstad** [comment]
> 1. What should be the name of the new non-owning resource?
>    `cuda_pool_wrapper`, `cuda_async_non_owning_memory_resource` , `cuda_async_shared_memory_resource`, `cuda_async_view_memory_resource`, ... ?

I like `cuda_pool_adaptor`.

> 2. Should the current `cuda_async_memory_resource` be able to take ownership of an existing pool, or should it always create its own pool ?

That would be difficult to do without an existing owning type for a `cudaPool_t`.

**Comment by harrism** [comment]
Had an offline discussion with @jrhemstad . I don't like the names `cuda_pool_adaptor` or `cuda_pool_wrapper`. My issue is as I mentioned before. I don't like having two MR types that basically work the same, but one owns the pool and one doesn't. I would prefer giving the two MRs similar names that reflect their similarity. So the suggestion is that we have:

* `cuda_async_memory_resource`: the existing pool-owning cudaMallocAsync MR.  
* `cuda_async_view_memory_resource`: the new pool-viewing (non-owning) cudaMallocAsync MR.

**Comment by harrism** [comment]
@fkallen any response?

**Comment by fkallen** [comment]
@harrism Sorry, I did not focus on this PR. Thanks for reminding me.
I agree with your view on the naming of the new class. I will change `cuda_pool_wrapper` to `cuda_async_view_memory_resource` .

I have more questions. At the moment I have added the constructor `cuda_async_memory_resource(cudaMemPool_t)` which takes ownership of the pool. Do we still need this if we have a dedicated view type? My original intention was to be able to use a raw cudaMemPool_t with rmm. Now, the view will be sufficient. 
Should there also be an additional constructor `cuda_async_memory_resource(cuda_async_view_memory_resource)` ?

**Comment by harrism** [comment]
I agree, I think the view is sufficient. No need for the constructor that takes ownership.

File: tests/mr/device/cuda_async_view_mr_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  poolProps.allocType = cudaMemAllocationTypePinned;
+  poolProps.location.id = rmm::detail::current_device().value();
+  poolProps.location.type = cudaMemLocationTypeDevice;
+
+  cudaMemPool_t memPool{};
+
+  RMM_CUDA_TRY(cudaMemPoolCreate(&memPool, &poolProps));
+
+  {
+
```
Comment (harrism) [inline tests/mr/device/cuda_async_view_mr_tests.cpp:54]:
```suggestion
```

File: tests/mr/device/cuda_async_view_mr_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    const auto pool_init_size{100};
+    cuda_async_view_mr mr{memPool};
+    void* ptr = mr.allocate(pool_init_size);
+    mr.deallocate(ptr, pool_init_size);
+    RMM_CUDA_TRY(cudaDeviceSynchronize());
+
```
Comment (harrism) [inline tests/mr/device/cuda_async_view_mr_tests.cpp:60]:
```suggestion
```

**Review by harrism — APPROVED** [review]
Looks great! Just a few unnecessary blank lines.

File: include/rmm/mr/device/cuda_async_memory_resource.hpp [hunk 3]
```diff
@@ -166,12 +170,13 @@ class cuda_async_memory_resource final : public device_memory_resource {
    *
    * @param p Pointer to be deallocated
    */
-  void do_deallocate(void* ptr, std::size_t, rmm::cuda_stream_view stream) override
+  void do_deallocate(void* ptr, std::size_t size, rmm::cuda_stream_view stream) override
   {
 #ifdef RMM_CUDA_MALLOC_ASYNC_SUPPORT
-    if (ptr != nullptr) { RMM_ASSERT_CUDA_SUCCESS(cudaFreeAsync(ptr, stream.value())); }
+    if (ptr != nullptr) { pool_.deallocate(ptr, size, stream); }
```
Comment (jrhemstad) [inline include/rmm/mr/device/cuda_async_memory_resource.hpp:176]: This is repeating the nullptr check twice.

**Review by jrhemstad — APPROVED** [review]

**Review by rongou — APPROVED** [review]

**Comment by harrism** [comment]
ok to test

**Comment by harrism** [comment]
@fkallen can you fix the style failures? Helps to run clang-format locally (e.g. enable format-on-save).

**Comment by fkallen** [comment]
I hope this fixes the issues. I could not use `AlignConsecutiveBitFields` and `AllowShortEnumsOnASingleLine`  with clang-format 10 on my machine.

**Comment by harrism** [comment]
@fkallen we've merged some major changes from @robertmaynard  which now load the symbols for cudaMallocAsync and related functions using `dlopen`. See #990 . Unfortunately this means your PR may need some changes, or at least to resolve conflicts. We only have two days before 22.04 code freeze. If you have time to try to get this working before then it may still make it, but we need to move fast, otherwise we need to slip to the next release.

**Comment by fkallen** [comment]
I have merged your changes. In tests/mr/device/cuda_async_view_mr_tests.cpp I have disabled a test because `cudaDeviceGetDefaultMemPool` is not available using the new mechanism. This should not be an issue since the view is effectively tested when testing the cuda_async_memory_resource.

**Comment by harrism** [comment]
> I have merged your changes. In tests/mr/device/cuda_async_view_mr_tests.cpp I have disabled a test because `cudaDeviceGetDefaultMemPool` is not available using the new mechanism. This should not be an issue since the view is effectively tested when testing the cuda_async_memory_resource.

Should probably add that API to the new mechanism and enable the test...

**Comment by harrism** [comment]
Thanks @fkallen !

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**